go/pkg/services/log_collector/server.go

```diff
@@ -15,37 +15,491 @@
+			podLogOptions := &v1.PodLogOptions{
+				SinceTime: prevLogTime,
+			}
+			*prevLogTime = metav1.Now()
+			restClientRequest := lcs.kubeClientSet.CoreV1().Pods(lcs.namespace).GetLogs(podName, podLogOptions)
+
+			// stream logs
+			restReadCloser, err := restClientRequest.Stream(ctx)
+			if err != nil {
+				lcs.Logger.WarnWithCtx(ctx, "Failed to get pod log read/closer", "runId", runId)
+				//return "", errors.Wrap(err, "Failed to get pod log read/closer", "runId", runId)
+			}
+
+			buf := new(bytes.Buffer)
+			_, err = io.Copy(buf, restReadCloser)
+			restReadCloser.Close() // nolint: errcheck
+			if err != nil {
+				lcs.Logger.WarnWithCtx(ctx, "Failed to copy information from log stream to buffer", "runId", runId)
+				//return "", errors.Wrap(err, "Failed to read pod logs", "runId", runId)
+			}
+
+			// write log contents to file
+			filePath := lcs.resolvePodLogFilePath(runId, podName)
+			if err := common.EnsureFileExists(filePath); err != nil {
+				lcs.Logger.ErrorWithCtx(ctx,
+					"Failed to create log file",
+					"runId", runId,
+					"filePath", filePath)
+			}
+			if err := lcs.writeToFile(ctx, filePath, buf.Bytes(), true); err != nil {
+				lcs.Logger.ErrorWithCtx(ctx,
+					"Failed to write log contents to file",
+					"runId", runId,
+					"podName", podName)
+			}
+
+		case <-stopChan:
+			lcs.Logger.DebugWithCtx(ctx,
+				"Removing item from state file",
+				"runId", runId,
+				"podName", podName)
+
+			// remove run id from state file
+			if err := lcs.removeItemFromInProgress(ctx, runId); err != nil {
+				lcs.Logger.WarnWithCtx(ctx, "Failed to remove run id from state file")
+			}
+
+			return
+		}
+	}
+}
+
+// addItemToInProgress adds an item to the `in_progress` list in the state file
+func (lcs *LogCollectorServer) addItemToInProgress(ctx context.Context, runId string, selector map[string]string) error {
+
+	stateFile, err := lcs.getState()
+	if err != nil {
+		return errors.Wrap(err, "Failed to get state file")
+	}
+
+	loggedItem := LogItem{
+		RunId:         runId,
+		LabelSelector: selector,
+	}
+
+	stateFile.InProgress = append(stateFile.InProgress, loggedItem)
+
+	if err := lcs.writeStateToFile(ctx, stateFile); err != nil {
+		return errors.Wrap(err, "Failed to write state file")
+	}
+
+	return nil
+}
+
+// removeItemFromInProgress removes an item from the `in_progress` list in the state file
+func (lcs *LogCollectorServer) removeItemFromInProgress(ctx context.Context, runId string) error {
+
+	// get state file
+	stateFile, err := lcs.getState()
+	if err != nil {
+		return errors.Wrap(err, "Failed to get state file")
+	}
+
+	// find and remove run id item
+	indexToDelete := 0
+	for i, loggedItem := range stateFile.InProgress {
+		if loggedItem.RunId == runId {
+			indexToDelete = i
+		}
+	}
+
+	// remove item from in progress slice
+	stateFile.InProgress[indexToDelete] = stateFile.InProgress[len(stateFile.InProgress)-1]
+	stateFile.InProgress = stateFile.InProgress[:len(stateFile.InProgress)-1]
```

Comment:
Not sure this implementation would scale easily with each removal requiring `O(n)`.
Maybe you should keep a `InProgress map[string]LogItem` where the key is the `runId`, so that both inserts and removals are `O(1)`